src/middleware/cloudWatch.ts

```diff
@@ -0,0 +1,59 @@
+// To add logs to cloudwatch
+// Custom middlware: https://github.com/middyjs/middy#configurable-middlewares
+
+import { getData, WrappedDocument } from "@govtechsg/open-attestation";
+import middy, { MiddlewareObj } from "@middy/core";
+import { NotarisationResult } from "src/functionHandlers/notarisePdt/handler";
+import { HealthCertDocument, Observation } from "src/types";
+import { logError, trace } from "./trace";
+
+// export const { trace } = getLogger("cloudWatchMiddleware");
+export type Request = Pick<middy.Request, "event" | "response">;
+
+// Custom middleware must return an obhect of { before?: fn, after?: fn, onError?: fn }
+// Note: Do not return a response as
+// this will totally stop the execution of successive middlewares in any phase (before, after, onError) and returns an early response
+// https://github.com/middyjs/middy#interrupt-middleware-execution-early
+
+export class CloudWatchMiddleware
+  implements Pick<MiddlewareObj, "before" | "after">
+{
+  provider = "";
+
+  before = async (req: Request): Promise<void> => {
+    const wrappedDocument = req.event
+      .body as WrappedDocument<HealthCertDocument>;
+    const data = getData(wrappedDocument);
+    const provider: string = data.issuers[0].name;
+    this.provider = provider;
+    trace(`provider ${provider} attempting to notarise pdt...`);
+  };
+
+  after = async (req: Request): Promise<void> => {
+    const { body } = req.response;
+    try {
+      const notarisationResult: NotarisationResult = JSON.parse(body);
+      const { notarisedDocument } = notarisationResult;
+      const data: HealthCertDocument = getData(notarisedDocument);
+      const observation = data.fhirBundle.entry.find(
+        (entr) => entr.resourceType === "Observation"
+      ) as Observation;
+      const testName = observation.code.coding[0].display;
+      const { provider } = this;
+      if (/art/i.test(testName)) {
+        trace(`${provider} successfully notarised pdt of type art`);
```

Comment:
You already have a cloudwatch insight query for such entry?

can imagine it will have various combination like

accredify successfully notarised pdt of type art
accredify successfully notarised pdt of type pcr

so the query u write need to account for `{provider}` - hard coz provider is dynamic text, so need group by this keyword from within a string of text. Not sure if it is possible to make this a json and let the key be part of cloudwatch fields?

as well as `successfully notarised pdt of type art` and `successfully notarised pdt of type pcr` - which i suppose is quite easy since is fixed text

Reply:
oh i just saw your description
```
fields @message
| parse @message /(?<provider>.*) attempting to notarise pdt/
| stats count(*) by provider```

nv try before this, but looks nice

Reply:
then i think can resolve this, and try in staging